File: app/src/main/java/com/kieran/winnipegbusbackend/StopTime.java
package com.kieran.winnipegbusbackend;

import android.support.annotation.NonNull;

import java.io.Serializable;
import java.util.Calendar;
import java.util.Date;
import java.util.Locale;

public class StopTime implements Comparable, Serializable {
    private static final double DUE_TIME = 0.5;
    private static final String DUE_STRING = "Due";
    private static final double EARLY_TIME = -0.5;
    private static final double LATE_TIME = 0.5;
    private static final String EARLY_TEXT = "Early";
    private static final String LATE_TEXT = "Late";
    private static final String OK_TEXT = "Ok";
    private static transient Calendar calendar = Calendar.getInstance();
    private long milliseconds;

    public StopTime() {
        milliseconds = System.currentTimeMillis();
    }

    public StopTime(Date date) {
        milliseconds = date.getTime();
    }

    public StopTime(long milliseconds) {
        this.milliseconds = milliseconds;
    }

    public static double timeBehindMinutes(StopTime estimated, StopTime scheduled) {
        return (estimated.milliseconds - scheduled.milliseconds) / 1000 / 60;
    }

    public static String getTimeStatus(StopTime estimated, StopTime scheduled) {
        double timeStatusNum = timeBehindMinutes(estimated, scheduled);

        if (timeStatusNum < LATE_TIME && timeStatusNum > EARLY_TIME)
            return OK_TEXT;
        else if (timeStatusNum <= EARLY_TIME)
            return EARLY_TEXT;
        else
            return LATE_TEXT;
    }

    public void increaseHour(int increase) {
        milliseconds += increase * 1000 * 60 * 60;
    }

    public String getMinutesString() {
        int minutes = getMinutes();
        return (minutes >= 10) ? String.valueOf(minutes) : ("0" + String.valueOf(minutes));
    }

    public String getHoursString() {
        return Integer.toString(getHours());
    }

    public String toString() {
        return getHoursString() + ":" + getMinutesString();
    }

    public long getMilliseconds() {
        return milliseconds;
    }

    public String to12hrTimeString() {
        String timeString;
        int hours = getHours();

        if(hours == 0)
            timeString = "12" + ":" + getMinutesString();
        else if(hours <= 12)
            timeString = getHoursString() + ":" + getMinutesString();
        else
            timeString = (hours - 12) + ":" + getMinutesString();

        timeString += (hours >= 12) ? "p" : "a";

        return timeString;
    }

    public String toFormattedString(StopTime currentTime, boolean use24hrTime) {
        if(currentTime != null) {
            double remainingTime = timeBehindMinutes(this, currentTime);

            if (remainingTime < DUE_TIME)
                return DUE_STRING;
            else if (remainingTime < 15)
                return Math.round(remainingTime) + " min";
        }

        if(use24hrTime)
            return this.to24hrTimeString();
        else
            return this.to12hrTimeString();
    }

    public String toFormattedDateString(boolean use24hrTime) {
        return  calendar.getDisplayName(Calendar.MONTH, Calendar.SHORT, Locale.CANADA) + " " + calendar.get(Calendar.DAY_OF_MONTH) + " " +  toFormattedString(null, use24hrTime);
    }

    public String toDateString() {
        String dateString = "";
        calendar.setTimeInMillis(milliseconds);
        int month = (calendar.get(Calendar.MONTH) + 1);
        int day = calendar.get(Calendar.DAY_OF_MONTH);

        dateString += calendar.get(Calendar.YEAR);
        dateString += "-" + (month >= 10 ? month : "0" + month);
        dateString += "-" + (day >= 10 ? day : "0" + day);

        return dateString;
    }

    public String toURLTimeString() {
        return toDateString() + "T" + to24hrTimeString() + ":00";
    }

    public String to24hrTimeString() {
        int hours = getHours();
        return ((hours < 10) ? "0" + getHoursString() : getHoursString()) + ":" + getMinutesString();
    }

    @Override
    public int compareTo(@NonNull Object another) {
        return (int)(milliseconds - ((StopTime)another).getMilliseconds());
    }

    public void decreaseMilliSeconds(long time) {
        milliseconds -= time;
    }

    private int getMinutes() {
        calendar.setTimeInMillis(milliseconds);
        return  calendar.get(Calendar.MINUTE);
    }

    private int getHours() {
        calendar.setTimeInMillis(milliseconds);
        return  calendar.get(Calendar.HOUR_OF_DAY);
    }

    public void increaseMinute(int minutes) {
        milliseconds += minutes * 1000 * 60;
    }
}

File: app/src/main/java/com/kieran/winnipegbusbackend/StopTime.kt
package com.kieran.winnipegbusbackend

import android.annotation.SuppressLint

import java.io.Serializable
import java.text.ParseException
import java.text.SimpleDateFormat
import java.util.Calendar
import java.util.Date
import java.util.Locale

class StopTime : Comparable<Any>, Serializable {
    var milliseconds: Long = 0
        private set

    val minutesString: String
        get() {
            val minutes = minutes
            return if (minutes >= 10) minutes.toString() else "0" + minutes.toString()
        }

    val hoursString: String
        get() = Integer.toString(hours)

    val minutes: Int
        get() {
            calendar.timeInMillis = milliseconds
            return calendar.get(Calendar.MINUTE)
        }

    val hours: Int
        get() {
            calendar.timeInMillis = milliseconds
            return calendar.get(Calendar.HOUR_OF_DAY)
        }

    val year: Int
        get() {
            calendar.timeInMillis = milliseconds
            return calendar.get(Calendar.YEAR)
        }

    val month: Int
        get() {
            calendar.timeInMillis = milliseconds
            return calendar.get(Calendar.MONTH)
        }

    val dayOfMonth: Int
        get() {
            calendar.timeInMillis = milliseconds
            return calendar.get(Calendar.DAY_OF_MONTH)
        }

    constructor() {
        milliseconds = System.currentTimeMillis()
    }

    constructor(date: Date) {
        milliseconds = date.time
    }

    constructor(milliseconds: Long) {
        this.milliseconds = milliseconds
    }

    fun increaseHour(increase: Int) {
        milliseconds += (increase * 1000 * 60 * 60).toLong()
    }

    override fun toString(): String {
        return hoursString + ":" + minutesString
    }

    fun to12hrTimeString(): String {
        var timeString: String
        val hours = hours

        if (hours == 0)
            timeString = "12" + ":" + minutesString
        else if (hours <= 12)
            timeString = hoursString + ":" + minutesString
        else
            timeString = (hours - 12).toString() + ":" + minutesString

        timeString += if (hours >= 12) "p" else "a"

        return timeString
    }

    fun toFormattedString(currentTime: StopTime?, use24hrTime: Boolean): String {
        if (currentTime != null) {
            val remainingTime = timeBehindMinutes(this, currentTime)

            if (remainingTime < DUE_TIME)
                return DUE_STRING
            else if (remainingTime < 15)
                return Math.round(remainingTime).toString() + " min"
        }

        return if (use24hrTime)
            this.to24hrTimeString()
        else
            this.to12hrTimeString()
    }

    fun toFormattedDateString(use24hrTime: Boolean): String {
        calendar.timeInMillis = milliseconds
        return calendar.getDisplayName(Calendar.MONTH, Calendar.SHORT, Locale.CANADA) + " " + calendar.get(Calendar.DAY_OF_MONTH) + " " + toFormattedString(null, use24hrTime)
    }

    fun toDateString(): String {
        var dateString = ""
        calendar.timeInMillis = milliseconds
        val month = calendar.get(Calendar.MONTH) + 1
        val day = calendar.get(Calendar.DAY_OF_MONTH)

        dateString += calendar.get(Calendar.YEAR)
        dateString += "-" + if (month >= 10) month else "0" + month
        dateString += "-" + if (day >= 10) day else "0" + day

        return dateString
    }

    fun toURLTimeString(): String {
        return toDateString() + "T" + to24hrTimeString() + ":00"
    }

    fun to24hrTimeString(): String {
        val hours = hours
        return (if (hours < 10) "0" + hoursString else hoursString) + ":" + minutesString
    }

    override operator fun compareTo(other: Any): Int {
        return (milliseconds - (other as StopTime).milliseconds).toInt()
    }

    fun decreaseMilliSeconds(time: Long) {
        milliseconds -= time
    }

    fun increaseMinute(minutes: Int) {
        milliseconds += (minutes * 1000 * 60).toLong()
    }

    fun toDatePickerDateFormat(): String {
        return datePickerDateFormat.format(Date(milliseconds))
    }

    companion object {
        private val DATE_FORMAT = "yyyy-MM-dd'T'HH:mm:ss"
        private val DUE_TIME = 0.5
        private val DUE_STRING = "Due"
        private val EARLY_TIME = -0.5
        private val LATE_TIME = 0.5
        private val EARLY_TEXT = "Early"
        private val LATE_TEXT = "Late"
        private val OK_TEXT = "Ok"
        @Transient private val calendar = Calendar.getInstance()
        private val datePickerDateFormat = SimpleDateFormat("EEE, MMM dd", Locale.CANADA)

        fun timeBehindMinutes(estimated: StopTime, scheduled: StopTime): Double {
            return ((estimated.milliseconds - scheduled.milliseconds) / 1000 / 60).toDouble()
        }

        fun getTimeStatus(estimated: StopTime, scheduled: StopTime): String {
            val timeStatusNum = timeBehindMinutes(estimated, scheduled)

            return if (timeStatusNum < LATE_TIME && timeStatusNum > EARLY_TIME)
                OK_TEXT
            else if (timeStatusNum <= EARLY_TIME)
                EARLY_TEXT
            else
                LATE_TEXT
        }

        fun convertStringToStopTime(s: String): StopTime? {
            @SuppressLint("SimpleDateFormat")
            val dateFormat = SimpleDateFormat(DATE_FORMAT)

            return try {
                StopTime(dateFormat.parse(s))
            } catch (e: ParseException) {
                null
            }

        }
    }
}
